File: YYReader/Parsing/HTMLParsingSupport.swift
import Foundation
import SwiftSoup

enum HTMLParsingSupport {
    static func document(from loaded: LoadedHTML) throws -> Document {
        try SwiftSoup.parse(loaded.html, loaded.finalURL.absoluteString)
    }

    static func absoluteURL(for element: Element, relativeTo baseURL: URL) -> URL? {
        guard let href = try? element.attr("href"),
              !href.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty,
              !href.lowercased().hasPrefix("javascript:") else {
            return nil
        }
        if let absolute = try? element.attr("abs:href"), !absolute.isEmpty,
           let url = URL(string: absolute) {
            return url
        }
        return URL(string: href, relativeTo: baseURL)?.absoluteURL
    }

    static func link(
        in document: Document,
        matching labels: Set<String>,
        baseURL: URL
    ) throws -> URL? {
        for anchor in try document.select("a").array() {
            let label = try anchor.text().trimmingCharacters(in: .whitespacesAndNewlines)
            if labels.contains(label), let url = absoluteURL(for: anchor, relativeTo: baseURL) {
                return url
            }
        }
        return nil
    }

    static func paragraphs(from element: Element) throws -> [String] {
        try element.select("script, style, noscript, iframe, form, .ad, .ads, .advertisement").remove()
        var source = try element.html()
        source = replacingRegex("(?i)<br\\s*/?>", in: source, with: "\n")
        source = replacingRegex("(?i)</(?:p|div|li|section|article|h[1-6])>", in: source, with: "\n")
        source = replacingRegex("(?i)<(?:p|div|li|section|article|h[1-6])[^>]*>", in: source, with: "")

        return source
            .split(separator: "\n", omittingEmptySubsequences: true)
            .compactMap { fragment -> String? in
                let text = (try? SwiftSoup.parseBodyFragment(String(fragment)).text()) ?? ""
                let cleaned = normalize(text)
                return cleaned.isEmpty ? nil : cleaned
            }
    }

    static func normalize(_ text: String) -> String {
        text
            .replacing("\u{00a0}", with: " ")
            .replacing("\u{3000}", with: " ")
            .split(whereSeparator: { $0.isWhitespace })
            .joined(separator: " ")
            .trimmingCharacters(in: .whitespacesAndNewlines)
    }

    static func replacingRegex(_ pattern: String, in value: String, with replacement: String) -> String {
        guard let regex = try? NSRegularExpression(pattern: pattern) else { return value }
        let range = NSRange(value.startIndex..., in: value)
        return regex.stringByReplacingMatches(in: value, range: range, withTemplate: replacement)
    }

    static func firstCapture(_ pattern: String, in value: String) -> String? {
        guard let regex = try? NSRegularExpression(pattern: pattern),
              let match = regex.firstMatch(in: value, range: NSRange(value.startIndex..., in: value)),
              match.numberOfRanges > 1,
              let range = Range(match.range(at: 1), in: value) else {
            return nil
        }
        return String(value[range])
    }

    static func chapterNumber(in title: String) -> Int? {
        guard let value = firstCapture("第\\s*(\\d+)\\s*[章回节]", in: title) else { return nil }
        return Int(value)
    }

    static func isSameOrigin(_ candidate: URL, as origin: URL) -> Bool {
        candidate.scheme?.lowercased() == origin.scheme?.lowercased()
            && candidate.host?.lowercased() == origin.host?.lowercased()
            && candidate.port == origin.port
    }
}

File: windows/YYReader.Windows.Core/Parsing/HtmlParsingSupport.cs
using System.Text;
using System.Text.RegularExpressions;
using AngleSharp.Dom;
using AngleSharp.Html.Parser;
using YYReader.Windows.Core.Models;

namespace YYReader.Windows.Core.Parsing;

public static class HtmlParsingSupport
{
    private static readonly HashSet<string> BlockTags = new(StringComparer.OrdinalIgnoreCase)
    {
        "article", "div", "h1", "h2", "h3", "h4", "h5", "h6", "li", "p", "section"
    };

    private static readonly HashSet<string> IgnoredTags = new(StringComparer.OrdinalIgnoreCase)
    {
        "form", "iframe", "noscript", "script", "style"
    };

    public static IDocument ParseDocument(LoadedHtml loaded) =>
        new HtmlParser().ParseDocument(loaded.Html);

    public static Uri? AbsoluteUrl(IElement element, Uri baseUri)
    {
        var href = element.GetAttribute("href")?.Trim();
        if (string.IsNullOrWhiteSpace(href)
            || href.StartsWith("javascript:", StringComparison.OrdinalIgnoreCase)
            || href.StartsWith("mailto:", StringComparison.OrdinalIgnoreCase)
            || href.StartsWith("data:", StringComparison.OrdinalIgnoreCase))
        {
            return null;
        }

        return Uri.TryCreate(baseUri, href, out var url) && url is not null
            ? UrlCanonicalizer.Canonicalize(url)
            : null;
    }

    public static Uri? Link(IDocument document, IEnumerable<string> labels, Uri baseUri)
    {
        var wanted = labels.ToHashSet(StringComparer.Ordinal);
        return document.QuerySelectorAll("a")
            .Select(anchor => (anchor, label: Normalize(anchor.TextContent)))
            .Where(pair => wanted.Contains(pair.label))
            .Select(pair => AbsoluteUrl(pair.anchor, baseUri))
            .FirstOrDefault(url => url is not null);
    }

    public static IReadOnlyList<string> Paragraphs(IElement element)
    {
        var paragraphs = new List<string>();
        var current = new StringBuilder();
        Walk(element, current, paragraphs);
        Flush(current, paragraphs);
        return paragraphs;
    }

    public static string Normalize(string text)
    {
        var value = text.Replace('\u00a0', ' ').Replace('\u3000', ' ');
        value = Regex.Replace(value, @"\s+", " ", RegexOptions.CultureInvariant);
        return value.Trim();
    }

    public static string? FirstCapture(string pattern, string value)
    {
        var match = Regex.Match(value, pattern, RegexOptions.CultureInvariant);
        return match.Success && match.Groups.Count > 1 ? match.Groups[1].Value : null;
    }

    public static int? ChapterNumber(string title)
    {
        var value = FirstCapture(@"第\s*(\d+)\s*[章回节]", title);
        return int.TryParse(value, out var number) ? number : null;
    }

    public static bool IsSameOrigin(Uri candidate, Uri origin) =>
        UrlCanonicalizer.IsSameOrigin(candidate, origin);

    private static void Walk(INode node, StringBuilder current, ICollection<string> paragraphs)
    {
        if (node is IText text)
        {
            current.Append(text.Data);
            return;
        }

        if (node is not IElement element)
        {
            return;
        }

        if (IgnoredTags.Contains(element.LocalName)
            || IsAdvertisement(element))
        {
            return;
        }

        if (element.LocalName.Equals("br", StringComparison.OrdinalIgnoreCase))
        {
            Flush(current, paragraphs);
            return;
        }

        foreach (var child in element.ChildNodes)
        {
            Walk(child, current, paragraphs);
        }

        if (BlockTags.Contains(element.LocalName))
        {
            Flush(current, paragraphs);
        }
    }

    private static bool IsAdvertisement(IElement element)
    {
        var id = element.Id;
        var classes = string.Join(' ', element.ClassList);
        return IsAdvertisementToken(id) || classes.Split(' ').Any(value => IsAdvertisementToken(value));
    }

    private static bool IsAdvertisementToken(string? value)
    {
        if (string.IsNullOrWhiteSpace(value))
        {
            return false;
        }

        var normalized = value.Trim().ToLowerInvariant();
        return normalized is "ad" or "ads" or "advertisement"
            || normalized.StartsWith("ad-", StringComparison.Ordinal)
            || normalized.StartsWith("ads-", StringComparison.Ordinal)
            || normalized.StartsWith("advertisement-", StringComparison.Ordinal);
    }

    private static void Flush(StringBuilder current, ICollection<string> paragraphs)
    {
        var value = Normalize(current.ToString());
        if (value.Length > 0)
        {
            paragraphs.Add(value);
        }

        current.Clear();
    }
}
